Mathlib/Tactic/Isolate/Core.lean

```diff
@@ -0,0 +1,213 @@
+/-
+Copyright (c) 2025 Heather Macbeth. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Heather Macbeth
+-/
+import Mathlib.Tactic.GCongr.Core
+import Mathlib.Control.Basic
+import Qq
+
+/-! # Tactic for isolating a subexpression within a given relation -/
+
+namespace Mathlib.Tactic.Isolate
+open Lean Meta Qq
+
+
+initialize registerTraceClass `Meta.isolate
+
+/-- Environment extension for "isolation" (`isolate`) lemmas. -/
+initialize isolateExt : SimpleScopedEnvExtension ((Name × Name × Nat × Bool) × Name)
+    (Std.HashMap (Name × Name × Nat × Bool) (Array Name)) ←
+  registerSimpleScopedEnvExtension {
+    addEntry := fun m (n, lem) => m.insert n ((m.getD n #[]).push lem)
+    initial := {}
+  }
+
+/-- Parse a lemma to see whether it has the correct form for a "isolation" (`isolate`) lemma. Such
+lemmas must have a conclusion of the form `f a₁ a₂ ... x ... aₖ ~ y ↔ x ~' G`; that is,
+* an if-and-only-if statement;
+* between two relations `~` and `~'`;
+* with a free variable `y` on one side of the relation `~` and a free variable `x` on one side of
+  the relation `~'`
+* and with a function application as the other side of the relation `~`, with `x` appearing exactly
```

Comment:
It took me a while to parse this, as "other side" referring back to the "one side with `y`".